### ISSUE TITLE
How to get Depth map image(jpg) of an JPG image in Android?

### ISSUE THREAD

**Issue by Istiakmorsalin** [dragon66/icafe#90]
I have used the libraries following code to get a depth map image:

`JPEGTweaker.extractDepthMap()`

But the problem is that the only sample pictures are providing outputs. Other images like pictures taken on pixel 2 xl or pixel 3(Even in the portrait mode) are not working.

**Comment by dragon66** [comment]
Can you provide sample images for me to see if possible?

**Comment by Istiakmorsalin** [comment]
@dragon66 Here are the images. I sent you these via email too. Please check. 

First one is a pixel 2XL device photo. 
![00100dPORTRAIT_00100_BURST20200218104958578_COVER](https://user-images.githubusercontent.com/2936695/74705227-5faedf00-523d-11ea-9b25-7a7cc383e501.jpg)

The second one is a bokeh image. 
![Josefina_with_Bokeh](https://user-images.githubusercontent.com/2936695/74705264-8240f800-523d-11ea-9770-ab5f37965ea1.jpg)

**Comment by dragon66** [comment]
The current implementation for extracting depth map is based on an earlier format of Google depth map documentation. Google has since changed to something called dynamic depth map.

I did some homework regarding the new Google dynamic depth map format and although it specifies the depth map data is saved under one data URI called android/depth map, I couldn't find it in either the image you provided or a similar image I downloaded from a post talking about similar issue.

The XMP metadata did contain quite a bit information regarding the depth map but not the map itself as in the old format did

**Comment by thyung** [comment]
Pixel phone portrait mode photo is concatenated of 4 JFIF structure [https://en.wikipedia.org/wiki/JPEG_File_Interchange_Format](https://en.wikipedia.org/wiki/JPEG_File_Interchange_Format) .  Each JFIF structure is an jpeg image.

A JFIF structure starts with marker 0xFFD8 and ends with marker 0xFFD9. Therefore, we can split a portrait mode image into 4 jpeg files.

The following python code prints the marker positions and splits PXL_20210102_093446326.PORTRAIT.jpg into, 

1. pxl_out_0.jpg: display image
2. pxl_out_1.jpg: original image
3. pxl_out_2.jpg: depthmap with 256 grey level
4. pxl_out_3.jpg: dummy image filled with 255

```python
with open('PXL_20210102_093446326.PORTRAIT.jpg', mode='rb') as infile:
    buffer = infile.read()

bufferlen = len(buffer)
pos = 0
pos_d8 = 0
n = 0
i = 0
while i < bufferlen:
    if buffer[i] == 0xff:
        pos = i
        i += 1
        if buffer[i] == 0xd8:
            print('ffd8: {0}'.format(pos))
            pos_d8 = pos
        elif buffer[i] == 0xd9:
            print('ffd9: {0} len: {1}'.format(pos, pos - pos_d8 + 2))
            with open('pxl_out_{0}.jpg'.format(n), mode='wb') as outfile:
                n += 1
                outfile.write(buffer[pos_d8: pos + 2])
    i += 1
```

Original image: PXL_20210102_093446326.PORTRAIT.jpg
![PXL_20210102_093446326 PORTRAIT](https://user-images.githubusercontent.com/6303254/104202712-3b1bb100-5466-11eb-82cb-6f62969e9c65.jpg)

Depthmap: pxl_out_2.jpg
![pxl_out_2](https://user-images.githubusercontent.com/6303254/104202880-6e5e4000-5466-11eb-81c3-be4a7209a09a.jpg)

**Comment by dragon66** [comment]
@thyung thanks for the information. So if I understand it correctly, basically the JPG image actually consists of 4 images conforming to JFIF spec and one of it is the depth map?

Wen

**Comment by thyung** [comment]
@dragon66 different phones have different ways to store depthmap.  My example photo was taken by Pixel 4.  I see a website tells iPhone depthmap is store in exif and can be extracted by

`exiftool -b -MPImage2 IMG_XXXX.jpg > IMG_XXXX_depth.jpg`

**Comment by tedca77** [comment]
I have been seeing GImage and GData (Depth map) within the XMP.  Being XMP/RDF/XML, it could be an XML element or could be an attribute.  I have been able to find the Depth map using an RDF library (RF4J) - you can look for large text fields containing the serialised image.  You could also process as XML, but you then have the issue of finding the field.  I can see it would be difficult to develop a generalised approach.

**Comment by dragon66** [comment]
Just finished developing an API to easily work with XMP metadata. Not sure that will help. Will come to this issue some time later.

**Comment by dragon66** [comment]
These Google Pixel phone images still use XMP data to store depthmap and other URIs but they don't store them as the sample image I used to extract depth map. The example image provided by @thyung does contain 4 normal images simply stacked together. I don't see any depth map images from the other two images provided by the OP.

One interesting thing regarding the image provided by @thyung is it actually hides anther 90 degree rotated color image a bit smaller than the actual showing one (the first image) in the XMP metadata section which is Base64 encoded. Similar to what my original extration does, I was able to extract it with a bit change of my code. However it is not a depth map.

Case by case it seems not that difficult but no general approach that will work for all with a simple method.

This along with the maker note section from Exif deserve a lot more attention.

I will keep this issue open in case I figure out a better, comprehensive approach to handle it.

**Comment by tedca77** [comment]
Would one approach to to provide via the API a list of all "large" text properties and their element names and get the user to choose which ones to extract. In my case, I am feeding the RDF into an RDF repository so I jremove any properties over a certain size.

**Comment by dragon66** [comment]
From an API perspective, I will not add such a method. This could be a post processing process on the XMP document ( the API could export RDF as binary or xml string).

One of the choice is SAX event driven approach to filter out the case where text node value is outside a limit. This is memory efficient.

Another approach is use DOM to filter out large text node. This will require simpler code but takes memory.

**Comment by tedca77** [comment]
> From an API perspective, I will not add such a method. 

Yes, that seems sensible - once a user has retrieved the XMP - they can do what they need with it.
